rust/src/types/stack.rs

```diff
@@ -1,15 +1,42 @@
-use std::cmp::min;
+use std::{cmp::min, sync::Mutex};
 
 use crate::types::{u256, FailStatus};
 
+static REUSABLE_STACK: Mutex<Option<Vec<u256>>> = Mutex::new(None);
```

Comment:
How does this perform with the CT? Where we have many instances of the processor running concurrently?

Reply:
If there is no allocation which can be reused because it is already in use, a new allocation is created. And when in Drop there is already one allocation in REUSABLE_STACK then one of those allocations (the one that was previously put there) will get deallocated. I will rewrite it to Mutex<Vec<Vec> and see how that performs...

Reply:
with Mutex<Vec<Vec<u256>>> which allows to reuse all allocations the performance is slightly worse in the (single threaded) benchmarks.

```
                   │ 2024-12-05T12:53#35e66b0#20-main/evmrs#performance │ 2024-12-05T13:12#30dd6c4#20/evmrs#performance │ 2024-12-05T14:44#1d02497#20/evmrs#performance │
                   │                       sec/op                       │         sec/op          vs base               │         sec/op          vs base               │
StaticOverhead/1/                                           1.891µ ± 1%              1.756µ ± 0%  -7.11% (p=0.000 n=20)              1.778µ ± 1%  -5.95% (p=0.000 n=20)
Inc/1/                                                      4.264µ ± 0%              4.133µ ± 0%  -3.08% (p=0.000 n=20)              4.277µ ± 0%       ~ (p=0.140 n=20)
Inc/10/                                                     4.243µ ± 1%              4.116µ ± 0%  -2.99% (p=0.000 n=20)              4.277µ ± 0%  +0.79% (p=0.002 n=20)
Fib/1/                                                      3.763µ ± 1%              3.691µ ± 0%  -1.91% (p=0.000 n=20)              3.824µ ± 1%  +1.63% (p=0.000 n=20)
Fib/5/                                                      12.60µ ± 1%              12.56µ ± 0%       ~ (p=0.133 n=20)              12.54µ ± 0%  -0.49% (p=0.020 n=20)
Fib/10/                                                     111.9µ ± 0%              113.4µ ± 0%  +1.30% (p=0.000 n=20)              111.6µ ± 0%  -0.24% (p=0.000 n=20)
Fib/15/                                                     1.109m ± 0%              1.117m ± 0%  +0.67% (p=0.000 n=20)              1.110m ± 0%       ~ (p=0.383 n=20)
Fib/20/                                                     12.00m ± 0%              12.00m ± 0%       ~ (p=0.461 n=20)              12.00m ± 0%       ~ (p=0.820 n=20)
Sha3/1/                                                     2.172µ ± 0%              2.063µ ± 1%  -5.02% (p=0.000 n=20)              2.087µ ± 1%  -3.96% (p=0.000 n=20)
Sha3/10/                                                    3.530µ ± 0%              3.422µ ± 0%  -3.06% (p=0.000 n=20)              3.442µ ± 0%  -2.49% (p=0.000 n=20)
Sha3/100/                                                   16.78µ ± 0%              16.98µ ± 1%  +1.16% (p=0.000 n=20)              16.94µ ± 2%       ~ (p=0.542 n=20)
Sha3/1000/                                                  177.6µ ± 0%              180.5µ ± 0%  +1.63% (p=0.000 n=20)              185.4µ ± 0%  +4.39% (p=0.000 n=20)
Arith/1/                                                    5.055µ ± 0%              4.905µ ± 0%  -2.98% (p=0.000 n=20)              4.903µ ± 1%  -3.01% (p=0.000 n=20)
Arith/10/                                                   10.50µ ± 1%              10.38µ ± 1%  -1.12% (p=0.006 n=20)              10.37µ ± 1%  -1.16% (p=0.000 n=20)
Arith/100/                                                  64.47µ ± 0%              63.67µ ± 0%  -1.24% (p=0.000 n=20)              64.00µ ± 0%  -0.72% (p=0.000 n=20)
Arith/280/                                                  192.9µ ± 0%              192.8µ ± 0%       ~ (p=0.134 n=20)              192.5µ ± 0%  -0.23% (p=0.000 n=20)
Memory/1/                                                   6.902µ ± 1%              6.845µ ± 0%  -0.83% (p=0.013 n=20)              7.102µ ± 1%  +2.91% (p=0.000 n=20)
Memory/10/                                                  12.46µ ± 1%              12.50µ ± 1%       ~ (p=0.534 n=20)              12.77µ ± 0%  +2.50% (p=0.000 n=20)
Memory/100/                                                 67.82µ ± 0%              67.72µ ± 0%       ~ (p=0.445 n=20)              67.73µ ± 0%       ~ (p=0.314 n=20)
Memory/1000/                                                593.4µ ± 0%              592.3µ ± 0%  -0.18% (p=0.000 n=20)              591.9µ ± 0%  -0.25% (p=0.000 n=20)
Memory/10000/                                               5.617m ± 0%              5.606m ± 0%  -0.19% (p=0.000 n=20)              5.609m ± 0%  -0.14% (p=0.002 n=20)
Analysis/jumpdest/                                          1.851µ ± 1%              1.777µ ± 1%  -4.00% (p=0.000 n=20)              1.780µ ± 1%  -3.84% (p=0.000 n=20)
Analysis/stop/                                              1.843µ ± 1%              1.771µ ± 0%  -3.91% (p=0.000 n=20)              1.777µ ± 1%  -3.58% (p=0.000 n=20)
Analysis/push1/                                             1.833µ ± 1%              1.764µ ± 0%  -3.76% (p=0.000 n=20)              1.770µ ± 1%  -3.46% (p=0.000 n=20)
Analysis/push32/                                            1.827µ ± 1%              1.768µ ± 0%  -3.20% (p=0.000 n=20)              1.787µ ± 0%  -2.16% (p=0.000 n=20)
geomean                                                     22.71µ                   22.35µ       -1.62%                             22.54µ       -0.76%
```

Reply:
I guess we can merge either of them, but the real value of this optimization comes when we do benchmark recursive calls.  I would use any of these two for now and keep in mind that this policy can be changed.